src/topology/sheaves/skyscraper.lean

```diff
@@ -0,0 +1,184 @@
+supported only at open sets contain `p₀`, i.e. `𝓕(U) = A` if `p₀ ∈ U` and `𝓕(U) = *` if `p₀ ∉ U`
+where `*` is a terminal object of `C`. In terms of stalks, `𝓕` is supported at all specializations
+of `p₀`, i.e. if `p₀ ⤳ x` then `𝓕ₓ ≅ A` and if `¬ p₀ ⤳ x` then `𝓕ₓ ≅ *`.
+
+## Main definitions
+
+* `skyscraper_presheaf`: `skyscraper_presheaf p₀ A` is the skyscraper presheaf at point `p₀` with
+  value `A`.
+* `skyscraper_sheaf`: the skyscraper presheaf satisfies the sheaf condition.
+
+## Main statements
+
+* `skyscraper_presheaf_stalk_of_specializes`: if `y ∈ closure {p₀}` then the stalk of
+  `skyscraper_presheaf p₀ A` at `y` is `A`.
+* `skyscraper_presheaf_stalk_of_not_specializes`: if `y ∉ closure {p₀}` then the stalk of
+  `skyscraper_presheaf p₀ A` at `y` is `*` the terminal object.
+
+TODO: generalize universe level when calculating stalks, after generalizing universe level of stalk.
+-/
+
+noncomputable theory
+
+section
+
+open topological_space
+open category_theory category_theory.limits
+open Top
+open opposite
+
+universes u v w
+
+variables {X : Top.{u}} (p₀ : X) {C : Type v} [category.{w} C] (S : C)
+variables [has_terminal C] [Π (U : opens X), decidable (p₀ ∈ U)]
+
+/--
+A skyscraper presheaf is a presheaf supported at a single point: if `p₀ ∈ X` is a specified
+point, then the skyscraper presheaf `𝓕` with value `A` is defined by `U ↦ A` if `p₀ ∈ U` and
+`U ↦ *` if `p₀ ∉ A` where `*` is some terminal object.
+-/
+@[simps] def skyscraper_presheaf : presheaf C X :=
+{ obj := λ U, if p₀ ∈ unop U then S else terminal C,
+  map := λ U V i, if h : p₀ ∈ unop V
+    then eq_to_hom $ by erw [if_pos h, if_pos (le_of_hom i.unop h)]
+    else ((if_neg h).symm.rec terminal_is_terminal).from _,
+  map_id' := λ U,
+  begin
+    split_ifs,
+    { apply eq_to_hom_refl },
+    { exact ((if_neg h).symm.rec terminal_is_terminal).hom_ext _ _ },
+  end,
+  map_comp' := λ U V W iVU iWV,
+  begin
+    by_cases hW : p₀ ∈ unop W,
+    { have hV : p₀ ∈ unop V := le_of_hom iWV.unop hW,
+      simp only [dif_pos hW, dif_pos hV, eq_to_hom_trans] },
+    { rw [dif_neg hW], apply ((if_neg hW).symm.rec terminal_is_terminal).hom_ext }
+  end }
+
+section
+
+variables {p₀}
+
+lemma skyscraper_presheaf_obj_of_mem {U : opens X} (h : p₀ ∈ U) :
+  (skyscraper_presheaf p₀ S).obj (op U) = S := if_pos h
+
+lemma skyscraper_presheaf_obj_of_not_mem {U : opens X} (h : p₀ ∉ U) :
+  (skyscraper_presheaf p₀ S).obj (op U) = terminal C := if_neg h
+
+end
+
+end
+
+section
+
+-- In this section, we calculate the stalks for skyscraper presheaves.
+-- We need to restrict universe level.
+
+open topological_space
+open category_theory category_theory.limits
+open Top
+open opposite
+
+universes u v
+
+variables {X : Top.{u}} (p₀ : X) {C : Type v} [category.{u} C] (S : C) [has_terminal C]
+
+private lemma mem_nhds_of_not_specializes {y : X} (h : ¬p₀ ⤳ y) :
+  ∃ (U : open_nhds y), p₀ ∉ U.1 :=
+by { contrapose! h, exact specializes_iff_forall_open.2 (λ s o h₁, h ⟨⟨s, o⟩, h₁⟩) }
```

Comment:
I don't think this should be private. I think this should go in `topology.inseparable`, and potentially be upgraded into an iff.

Reply:
I have added 
```lean
lemma not_specializes_iff_exists_open : ¬ x ⤳ y  ↔ ∃ (S : set X), is_open S ∧ y ∈ S ∧ x ∉ S
lemma not_specializes_iff_exists_closed : ¬ x ⤳ y ↔ ∃ (S : set X), is_closed S ∧ x ∈ S ∧ y ∉ S :=
```
and removed the private lemma